sled-agent/src/bootstrap/rss_handle.rs

```diff
@@ -0,0 +1,232 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! sled-agent's handle to the Rack Setup Service it spawns
+
+use super::client as bootstrap_agent_client;
+use super::discovery::PeerMonitorObserver;
+use super::params::SledAgentRequest;
+use crate::rack_setup::config::SetupServiceConfig;
+use crate::rack_setup::service::Service;
+use futures::stream::FuturesUnordered;
+use futures::Future;
+use futures::StreamExt;
+use omicron_common::backoff::internal_service_policy;
+use omicron_common::backoff::retry_notify;
+use omicron_common::backoff::BackoffError;
+use slog::Logger;
+use std::net::SocketAddrV6;
+use thiserror::Error;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+
+pub(super) struct RssHandle {
+    _rss: Service,
+    task: JoinHandle<()>,
+}
+
+impl Drop for RssHandle {
+    fn drop(&mut self) {
+        // NOTE: Ideally, with async drop, we'd await completion of the our task
+        // handler.
+        //
+        // Without that option, we instead opt to simply cancel the task to
+        // ensure it does not remain alive beyond the handle itself.
+        self.task.abort();
+    }
+}
+
+impl RssHandle {
+    // Start the Rack setup service.
+    pub(super) fn start_rss(
+        log: &Logger,
+        config: SetupServiceConfig,
+        peer_monitor: PeerMonitorObserver,
+    ) -> Self {
+        let (tx, rx) = rss_sled_agent_channel();
+
+        let rss = Service::new(
+            log.new(o!("component" => "RSS")),
+            config,
+            peer_monitor,
+            tx,
+        );
+        let log = log.new(o!("component" => "BootstrapAgentRssHandler"));
+        let task = tokio::spawn(async move {
+            handle_rss_requests(&log, rx).await;
```

Comment:
This looks like it executes once, rather than in a loop, no? Is that intentional?

What if the sender only sent half the requests, and then we context switched to this receiver task? It seems like the latter half of requests would never be processed

Reply:
Follow-up: I see that we send a Vec of addresses in a single request, so the "cutoff" behavior doesn't seem possible. However, if the intent is that the sender can only be used once, we should document that behavior very explicitly

Reply:
Good call, and it would be better to enforce it statically. I think `RssSledAgentSender::initialize_sleds()` should be able to take `self` instead of `&self`, guaranteeing RSS can only call it once.

Reply:
This should be clarified (both in comments and in the type system) in 92aa2e2a